### PR TITLE
Change metrics func to get a stats reporter and propagate to the tracer

### PR DESCRIPTION
**Pull request by madhuravi**
Tally API upgrade removed access to the stats reporter from the scope. Tracer needs access to the stats reporter. This is to propagate and store stats reporter in service.Host so it can be used by the tracer.

### CODE REVIEW

**Comment by madhuravi** [comment]
@uber-go/service-framework

File: metrics/metrics.go [hunk 1]
```diff
@@ -72,33 +73,33 @@ func ensureNotFrozen() {
...
 
-	if _scopeFunc != nil {
-		rep, closer, err := _scopeFunc(i)
+	if _repFunc != nil {
+		rep, err := _repFunc(i)
 		if err != nil {
 			panic(fmt.Sprintf("Failed to initialize metrics reporter %v", err))
 		}
-		return rep, closer
+		scope, closer := tally.NewRootScope("", nil, rep, time.Second)
```
Comment (glibsm) [inline metrics/metrics.go]: This is not what we want, as I was trying to explain earlier. We want to give control over creating this scope to the service owners, if they choose to do so.

The parameters to this function is something that has to be configurable: prefix, tags, and the tick duration.

This is how this file initially looked, by taking in a reporter, but it's a lot less flexible than allowing service owners to create their own metrics structure that might work better for them. We can't make those sort of decisions (especially not the tags they want to include on all metrics events) for them.

**Comment by glibsm** [comment]
See #7 for more details

File: metrics/metrics.go (hunk 1 above)
Comment (madhuravi) [inline metrics/metrics.go]: Service owners can still configure using WithMetrics options. From further conversation with Glib, he suggested we might want to add special configuration for uber internal services. For this reason, changed to allowing scope and reporter to be passed in and use the pre-configured metrics scope.

File: metrics/metrics.go [hunk 2]
```diff
@@ -84,21 +83,20 @@ func RegisterRootScope(rep ScopeFunc) {
...
+		scope, reporter, closer, err := _scopeFunc(i)
 		if err != nil {
 			panic(fmt.Sprintf("Failed to initialize metrics reporter %v", err))
 		}
-		return rep, closer
+		return scope, reporter, closer
 	}
-
-	return nil, nil
+	return nil, nil, nil
```
Comment (anuptalwalkar) [inline metrics/metrics.go]: This is a strange func call to deal with. We should wrap the return values in a struct, and return it with possible error as second parameter.

Comment (madhuravi) [inline metrics/metrics.go]: What makes it strange? In other words, what's the advantage of wrapping this in a struct? We treat the error as a panic here, why would we want to return the error?

Comment (anuptalwalkar) [inline metrics/metrics.go]: Well, I am talking about dealing with individual return parameters is strange for the caller. Right now you have to check for each return variable in service.go and assign null scope, add check for each nil/not nil, and so on. encapsulating what rootScope returns would make the code concise and readable. 

For error I am not talking about panic, but instead of returning nil, nil, nil, it would make more sense for the caller to simply check for err message (something along the lines that no scope func is registered) 
The method would look like this - 
`
func RootScope(i ScopeInit) (scopeRegistry, error) {
  if _scopeFunc != nil {
    return scopeRegistry{}, nil
  }
  return nil, errors.New("no scope functions registered")
}
`

Comment (anuptalwalkar) [inline metrics/metrics.go]: I am not good with naming, but you can call scopeRegistry encapsulation rootScope or just scope..

File: service/host_test.go [hunk 3]
```diff
@@ -195,7 +195,7 @@ func TestHostShutdown_RunningService(t *testing.T) {
 func TestHostShutdown_CloseSuccessful(t *testing.T) {
 	sh := makeRunningHost()
 	sh.serviceCore.metricsCore = metricsCore{
-		scope:            tally.NoopScope,
+		metrics:          tally.NoopScope,
```
Comment (glibsm) [inline service/host_test.go]: nit: not sure i like renaming metrics to scope. Yes, we expose this publicly through `Metrics()` API, but internally we don't need to mix and match things, i.e. `call it like you see it`. It's a tally scope.

File: service/host_mock.go [hunk 4]
```diff
@@ -41,7 +41,7 @@ func NullHost() Host {
 			log: ulog.NoopLogger,
 		},
 		metricsCore: metricsCore{
-			scope: tally.NoopScope,
+			metrics: tally.NoopScope,
```
Comment (glibsm) [inline service/host_mock.go]: add a null stats reporter here as well, please otherwise some tests might crash due to runtime metrics trying to use it

File: metrics/metrics.go (hunk 2 above)
Comment (madhuravi) [inline metrics/metrics.go]: I don't think a scope registry is necessary here. This exchange happens literally in one place and the caller (service.go) assigns it to a struct. Creating a new struct seems redundant. Go has strongly typed multiple return values, so the caller knows exactly what to expect. 
With errors, if this is not a panic, I completely agree. There is no use for the returned error if we are handling it within.

File: service/host_mock.go (hunk 4 above)
Comment (madhuravi) [inline service/host_mock.go]: Won't crash because of the noop tracer but added it anyway.

File: service/host_test.go (hunk 3 above)
Comment (madhuravi) [inline service/host_test.go]: I see what you mean. The reason I finally went for this is if you look at service, we'll have
if svc.Metrics() == nil {
  svc.scope = tally.Scope
}

Checking Metrics() and setting scope appears unrelated to someone reading it. That's why I decided to keep them uniform.
